phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -945,20 +960,40 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
                     PDataType dataType = MetaDataUtil.getViewIndexIdDataType();
                     ColumnName colName = ColumnName.caseSensitiveColumnName(MetaDataUtil.getViewIndexIdColumnName());
                     allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, SortOrder.getDefault()));
-                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault()));
+                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault(), null));
                 }
                 // First columns are the indexed ones
-                for (Pair<ColumnName, SortOrder> pair : indexedPkColumns) {
-                    ColumnName colName = pair.getFirst();
-                    PColumn col = resolver.resolveColumn(null, colName.getFamilyName(), colName.getColumnName()).getColumn();
-                    unusedPkColumns.remove(col);
-                    // Ignore view constants for updatable views as we don't need these in the index
-                    if (col.getViewConstant() == null) {
-                        PDataType dataType = IndexUtil.getIndexColumnDataType(col);
-                        colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
-                        allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, pair.getSecond()));
-                        columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), col.isNullable(), col.getMaxLength(), col.getScale(), false, SortOrder.getDefault()));
-                    }
+                for (Pair<ParseNode, SortOrder> pair : indexParseNodeAndSortOrderList) {
+                	ParseNode parseNode = pair.getFirst();
+                	ColumnName colName = null;
+                	// if it is a column 
+                	if (parseNode instanceof ColumnParseNode) {
```

Comment:
Or a better way is to change unusedPkColumns to be a Set<Expression>. These can be created directly through

    List<PColumn> pkColumns = table.getPKColumns();
    List<Expression> unusedPkColumns = new LinkedHashSet(pkColumns.size());
    for (int i = 0; i < pkColumns; i++) {
        unusedPkColumns.add(new RowKeyColumnExpression(pkColumns.get(i), 
            new RowKeyValueAccessor(pkColumns, i));
    }

Then, here you can get rid of the special case, and instead after compilation, do the following:

    if (expression.isStateless()) { // Will be true for any constant (including a view constant)
        continue;
    }
    unusedPkColumns.remove(expression);

We don't need to include an expression that evaluates to a constant in the index. You can test this by adding a function index on something like 1 + 2.

In this way, you can remove the if statement completely.